### DISCUSSION TITLE
Lifecycle Event Bus

### DISCUSSION THREAD

**Discussion by waldronmatt** [lit/lit#3286] [Ideas]
On rare occasions I've needed to access the renderRoot of child components nested inside a parent shadowRoot. However, using the parent component's lifecycle methods doesn't guarantee access to child renderRoots - which I think is similar to this issue here https://github.com/WICG/webcomponents/issues/551

To overcome this at my workplace, I created a simple mixin imported in every component that emits a custom event when one of the lifecycle events trigger. This has been proving great, specifically in situations where I have a custom app that wants to do something when a child component is connected to the DOM or even when an update is complete. Additionally, this has been reliable for guaranteeing access to renderRoot.

I'd like to hear the Lit team's thoughts on this, as I didn't see this approach documented. Is this a good/bad approach and is this something the team has considered? Much appreciated.

mixin.ts

```ts
 class LitBase extends constructor {
    connectedCallback(): void {
      super.connectedCallback();

      const connectedCallbackEvent = new CustomEvent(
        'connected-callback', 
        { bubbles: true, composed: true }
      );
      this.dispatchEvent(connectedCallbackEvent);
    }

    ....

export class LitWrapper extends LitBaseMixin(LitElement) {}
```

child.ts
```ts
export class ChildComponent extends LitWrapper {
...
}
```

parent.ts
```ts
protected render() {
    return html`
      <child-component
          @connected-callback=${(event: { target: HTMLElement }) => {
            // do something here like accessing the shadowRoot
          }}>hello
      </child-component>` ;
    }
```

**Comment by justinfagnani** [comment]
Emitting events for every lifecycle is going to be expensive and slow down rendering.

Hopefully you don't need to access a child's shadowRoot that often, and when you do you can frame this as an API on the child component. Like `getSomeInternalElement(): Promise<HTMLElement>`. Then you could use `ref()` or a query in `firectUpdated` to get to it.

**Reply by waldronmatt** [reply]
Thanks Justin, good to know there is performance implications and concerns. You're right about firstUpdated method which has proved successful to gain entry to nested component shadowRoots. I opted for the approch above because i thought it was better but now I may opt back to firstUpdated. Interested in trying ref
